go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+// Copyright 2019 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package blobstore
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"math"
+	"os"
+	"path"
+	"sort"
+
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/oracle/oci-go-sdk/v65/objectstorage"
+	"golang.org/x/sync/errgroup"
+)
+
+// 10MB part size
+const minPartSize = 10 * 1024 * 1024
+const defaultPartSize = 5 * minPartSize
+const maxPartNum = 10000
+const defaultBatchSize = 500 * 1024 * 1024
+const defaultConcurrentListeners = 5
+
+type toUpload struct {
+	b       []byte
+	partNum int
+}
+
+type toDownload struct {
+	path    string
+	partNum int
+	br      BlobRange
+}
+
+type uploadFunc func(ctx context.Context, objectName, uploadID string, partNumber int, contentLength int64, reader io.Reader) (objectstorage.CommitMultipartUploadPartDetails, error)
+
+type downloadFunc func(ctx context.Context, objectName, namespace, etag string, partNumber int, br BlobRange) (*toDownload, error)
+
+type tempLocalObject struct {
+	path string
+	f    *os.File
+}
+
+var _ io.ReadCloser = &tempLocalObject{}
+
+func (t *tempLocalObject) Read(p []byte) (int, error) {
+	return t.f.Read(p)
+}
+
+func (t *tempLocalObject) Close() error {
+	err := t.f.Close()
+	os.Remove(t.path)
+	return err
+}
+
+// OCIBlobstore provides an OCI implementation of the Blobstore interface
+type OCIBlobstore struct {
+	provider            common.ConfigurationProvider
+	client              objectstorage.ObjectStorageClient
+	bucketName          string
+	nameSpace           string
+	prefix              string
+	concurrentListeners int
+}
+
+var _ Blobstore = &OCIBlobstore{}
+
+// NewOCIBlobstore creates a new instance of a OCIBlobstore
+func NewOCIBlobstore(provider common.ConfigurationProvider, client objectstorage.ObjectStorageClient, bucketName, prefix string) *OCIBlobstore {
+	for len(prefix) > 0 && prefix[0] == '/' {
+		prefix = prefix[1:]
+	}
+	return &OCIBlobstore{provider, client, bucketName, "", prefix, defaultConcurrentListeners}
+}
+
+func (bs *OCIBlobstore) Path() string {
+	return path.Join(bs.bucketName, bs.prefix)
+}
+
+func (bs *OCIBlobstore) getNamespace(ctx context.Context) (string, error) {
+	if bs.nameSpace != "" {
+		return bs.nameSpace, nil
+	}
+	request := objectstorage.GetNamespaceRequest{}
+	r, err := bs.client.GetNamespace(ctx, request)
+	if err != nil {
+		return "", err
+	}
+	bs.nameSpace = *r.Value
+	return bs.nameSpace, nil
+}
```

Comment:
This makes `bs` non-intuitively non-thread-safe. Can we do this at construction time instead?

Reply:
Or put it behind a `(*sync.Once).Do()` I suppose...